integreat_compass/cms/views/offers/offer_form_view.py

```diff
@@ -0,0 +1,178 @@
+import logging
+
+from django.contrib import messages
+from django.core.exceptions import PermissionDenied
+from django.db import transaction
+from django.shortcuts import render
+from django.utils.decorators import method_decorator
+from django.utils.translation import gettext as _
+from django.views.generic import TemplateView
+
+from ...decorators import permission_required
+from ...forms import (
+    ContactForm,
+    LocationForm,
+    OfferForm,
+    OfferVersionForm,
+    OrganizationForm,
+)
+from ...models import Offer, User
+
+logger = logging.getLogger(__name__)
+
+
+@method_decorator(permission_required("cms.edit_offer"), name="dispatch")
+class OfferFormView(TemplateView):
+    """
+    View for the offer form
+    """
+
+    template_name = "offers/offer_form.html"
+
+    def get(self, request, *args, **kwargs):
+        r"""
+        Render :class:`~integreat_compass.cms.forms.offers.contact_form.ContactForm`,
+               :class:`~integreat_compass.cms.forms.offers.location_form.LocationForm`,
+               :class:`~integreat_compass.cms.forms.offers.offer_form.OfferForm`,
+               :class:`~integreat_compass.cms.forms.offers.offer_version_form.OfferVersionForm` and
+               :class:`~integreat_compass.cms.forms.offers.organization_form.OrganizationForm`
+
+        :param request: The current request
+        :type request: ~django.http.HttpRequest
+
+        :param \*args: The supplied arguments
+        :type \*args: list
+
+        :param \**kwargs: The supplied keyword arguments
+        :type \**kwargs: dict
+
+        :raises ~django.core.exceptions.PermissionDenied: If user does not have the permission to edit the specific page
+
+        :return: The rendered template response
+        :rtype: ~django.template.response.TemplateResponse
+        """
+
+        offer = Offer.objects.filter(id=kwargs.get("offer_id")).first()
+        (offer_version, offer_contact, offer_location, offer_organization) = (
+            (
+                offer.versions.latest(),
+                offer.offer_contact,
+                offer.location,
+                offer.organization,
+            )
+            if offer is not None
+            else (None, None, None, None)
+        )
+        offer_form = OfferForm(instance=offer)
+        offer_version_form = OfferVersionForm(instance=offer_version)
+        contact_form = ContactForm(instance=offer_contact)
+        location_form = LocationForm(instance=offer_location)
+        organization_form = OrganizationForm(instance=offer_organization)
```

Comment:
I mean, I could add something like `_get_get_forms()` and `_get_post_forms()` of course, which would make the existing methods shorter, but does not really reduce code duplication 😅

Reply:
I didn't try this out, but imagined something along the lines of:
```
    def get_form_classes(self):
        return [OfferForm, LocationForm, OfferForm, OfferVersionForm, OrganizationForm]

    def get_form_kwargs(self, form_class):
        kwargs = {}
        if self.request.method in ["POST", "PUT"]:
            kwargs["data"] = self.request.POST
        if hasattr(self, "object"):
            kwargs["instance"] = form_class.get_instance_from_offer(self.object)
        return kwargs

    def get_forms(self):
        return [
            form_class(**self.get_form_kwargs(form_class))
            for form_class in self.get_form_classes()
        ]

    def get_context_data(self, **kwargs):
        for form in self.get_forms():
            kwargs[form.context_name] = form
        return super().get_context_data(**kwargs)
```

This would require to implement a `get_instance_from_offer()` method in each form class, and an attribute `context_name` which denotes the name of the variable to access the form in the template...

Not sure however if this really makes things easier or just more complicated :sweat_smile:

Reply:
It's a nice idea, but introducing (essentially) 9 new methods to avoid 5 lines not-quite-duplicate lines of code is probably excessive 😅

Reply:
Yes you're right :see_no_evil: 
Maybe if we happen to add more views with multiple forms :grin: